## Assignment Overview

### About This Assignment
In this assignment you will set up a professional Python development environment using Conda, install the four core ML libraries covered in class, and apply each library to real emergency call data from Montgomery County, Pennsylvania. This mirrors how data scientists work in the real world — environment first, code second.


### Step 1: Load Data



In [75]:
# Import all the necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [76]:
# load the csv file and store it in a dataframe called df
df = pd.read_csv('911.csv')

### 2. Exploratory Data Analysis

In [92]:
# Check the first 5 rows of the dataframe
df.head()             

,lat,lng,desc,zip,title,timeStamp,twp,addr,e,dept,reason
0,40.297876,-75.581294,REINDEER CT & DEAD END; NEW HANOVER; Station ...,19525.0,EMS: BACK PAINS/INJURY,2015-12-10 17:10:52,NEW HANOVER,REINDEER CT & DEAD END,1,EMS,BACK PAINS/INJURY
1,40.258061,-75.264680,BRIAR PATH & WHITEMARSH LN; HATFIELD TOWNSHIP...,19446.0,EMS: DIABETIC EMERGENCY,2015-12-10 17:29:21,HATFIELD TOWNSHIP,BRIAR PATH & WHITEMARSH LN,1,EMS,DIABETIC EMERGENCY
2,40.121182,-75.351975,HAWS AVE; NORRISTOWN; 2015-12-10 @ 14:39:21-St...,19401.0,Fire: GAS-ODOR/LEAK,2015-12-10 14:39:21,NORRISTOWN,HAWS AVE,1,Fire,GAS-ODOR/LEAK
3,40.116153,-75.343513,AIRY ST & SWEDE ST; NORRISTOWN; Station 308A;...,19401.0,EMS: CARDIAC EMERGENCY,2015-12-10 16:47:36,NORRISTOWN,AIRY ST & SWEDE ST,1,EMS,CARDIAC EMERGENCY
4,40.251492,-75.603350,CHERRYWOOD CT & DEAD END; LOWER POTTSGROVE; S...,NaN,EMS: DIZZINESS,2015-12-10 16:56:52,LOWER POTTSGROVE,CHERRYWOOD CT & DEAD END,1,EMS,DIZZINESS


In [78]:
# check the last 5 rows of the dataframe
df.tail()          

,lat,lng,desc,zip,title,timeStamp,twp,addr,e
663517,40.157956,-75.348060,SUNSET AVE & WOODLAND AVE; EAST NORRITON; 2020...,19403.0,Traffic: VEHICLE ACCIDENT -,2020-07-29 15:46:51,EAST NORRITON,SUNSET AVE & WOODLAND AVE,1
663518,40.136306,-75.428697,EAGLEVILLE RD & BUNTING CIR; LOWER PROVIDENCE...,19403.0,EMS: GENERAL WEAKNESS,2020-07-29 15:52:19,LOWER PROVIDENCE,EAGLEVILLE RD & BUNTING CIR,1
663519,40.013779,-75.300835,HAVERFORD STATION RD; LOWER MERION; Station 3...,19041.0,EMS: VEHICLE ACCIDENT,2020-07-29 15:52:52,LOWER MERION,HAVERFORD STATION RD,1
663520,40.121603,-75.351437,MARSHALL ST & HAWS AVE; NORRISTOWN; 2020-07-29...,19401.0,Fire: BUILDING FIRE,2020-07-29 15:54:08,NORRISTOWN,MARSHALL ST & HAWS AVE,1
663521,40.015046,-75.299674,HAVERFORD STATION RD & W MONTGOMERY AVE; LOWER...,19041.0,Traffic: VEHICLE ACCIDENT -,2020-07-29 15:52:46,LOWER MERION,HAVERFORD STATION RD & W MONTGOMERY AVE,1


In [79]:
# Check the number of rows and columns in the dataframe 
df.shape           

(663522, 9)

In [80]:
# check the information about the dataframe
df.info()              

<class 'pandas.DataFrame'>
RangeIndex: 663522 entries, 0 to 663521
Data columns (total 9 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   lat        663522 non-null  float64
 1   lng        663522 non-null  float64
 2   desc       663522 non-null  str    
 3   zip        583323 non-null  float64
 4   title      663522 non-null  str    
 5   timeStamp  663522 non-null  str    
 6   twp        663229 non-null  str    
 7   addr       663522 non-null  str    
 8   e          663522 non-null  int64  
dtypes: float64(3), int64(1), str(5)
memory usage: 45.6 MB


In [81]:
# Check the statisitcs summary
df.describe()

,lat,lng,zip,e
count,663522.000000,663522.000000,583323.000000,663522.0
mean,40.158162,-75.300105,19236.055791,1.0
std,0.220641,1.672884,298.222637,0.0
min,0.000000,-119.698206,1104.000000,1.0
25%,40.100344,-75.392735,19038.000000,1.0
50%,40.143927,-75.305143,19401.000000,1.0
75%,40.229008,-75.211865,19446.000000,1.0
max,51.335390,87.854975,77316.000000,1.0


In [82]:
# We can also transform the rows and columns of our statisitcal table
df.describe().T

,count,mean,std,min,25%,50%,75%,max
lat,663522.0,40.158162,0.220641,0.000000,40.100344,40.143927,40.229008,51.335390
lng,663522.0,-75.300105,1.672884,-119.698206,-75.392735,-75.305143,-75.211865,87.854975
zip,583323.0,19236.055791,298.222637,1104.000000,19038.000000,19401.000000,19446.000000,77316.000000
e,663522.0,1.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000


In [83]:
# Create new categories "dept" and "reason" from the "title" column 
df['dept'] = df['title'].str.split(':').str[0] 
df['reason'] = df['title'].str.split(':').str[1]

# Run value_count() on both of them.
df[['dept', 'reason']].value_counts()           # We are combining both of them


dept     reason               
Traffic  VEHICLE ACCIDENT -       148372
         DISABLED VEHICLE -        47909
Fire     FIRE ALARM                38336
EMS      FALL VICTIM               34676
         RESPIRATORY EMERGENCY     34248
                                   ...  
Fire     FOOT PATROL                   1
         DIZZINESS                     1
         HEMORRHAGING                  1
         CVA/STROKE                    1
         BARRICADED SUBJECT            1
Name: count, Length: 148, dtype: int64

In [84]:
# We can also group reason by dept
df.groupby('dept')['reason'].value_counts()

dept     reason                     
EMS      FALL VICTIM                    34676
         RESPIRATORY EMERGENCY          34248
         CARDIAC EMERGENCY              32332
         VEHICLE ACCIDENT               25513
         SUBJECT IN PAIN                19646
                                        ...  
Traffic  ROAD OBSTRUCTION -             23235
         HAZARDOUS ROAD CONDITIONS -     6833
         VEHICLE FIRE -                  3366
         VEHICLE LEAKING FUEL -           292
         DEBRIS/FLUIDS ON HIGHWAY -       201
Name: count, Length: 148, dtype: int64

In [ ]:
# Let's visualize
df['reason'].value_counts().head(10).plot(kind='barh', xlabel='frequency')


In [ ]:
# Visuale dept
df['dept'].value_counts().plot(kind='barh')

### Step 3:  Clean the dataset




In [ ]:
# Check the number of missing values in each column
df.isnull().sum()   

In [ ]:
# Check percentage of rows missing
df.isnull().sum() / len(df) * 100

We can see that there are missing values under columns "zip" and "twp" which is a a total of 80492 rows out of thr 663522 rows (about 12% of our rows). We would fill these rows intelligently instead of dropping them

In [ ]:
# Fill less critical categorical columns
df['twp'] = df['twp'].fillna('Unknown')
df['zip'] = df['zip'].fillna(df['zip'].mode()[0])  

In [ ]:
# Inspect duplicates
print(f" Duplicated rows: {int(df.duplicated().sum())}")

In [ ]:
# Check if the missing values have been filled
print(f" Shape after filling rows: {df.shape}")

In [ ]:
# Smart duplicate definition
duplicates = int(df.duplicated(subset=['lat', 'lng', 'timeStamp', 'title']).sum())
duplicates 

In [ ]:
# Convert time to datetime
df['timeStamp'] = pd.to_datetime(df['timeStamp'])

### Step 3: Check for Outliers

We are going to check for strange coordinates (lag, lng)


In [87]:
# Inspect range
df[['lat', 'lng']].describe()

,lat,lng
count,663522.000000,663522.000000
mean,40.158162,-75.300105
std,0.220641,1.672884
min,0.000000,-119.698206
25%,40.100344,-75.392735
50%,40.143927,-75.305143
75%,40.229008,-75.211865
max,51.335390,87.854975


For this database (Pennsylvanina area), roughly expect:

- **lat**: ~39.72° to 42.27°
- **lng**: ~-74.68° to -80.52°

In [88]:
# Find suspicious values
df[(df['lat'] < 39.72) | (df['lat'] > 42.27)]
df[(df['lng'] < -80.52) | (df['lng'] > -74.68)]


,lat,lng,desc,zip,title,timeStamp,twp,addr,e,dept,reason
25529,32.387090,-86.276106,600 ; MONTGOMERY; 2016-02-12 @ 18:44:37;,36107.0,Traffic: DISABLED VEHICLE -,2016-02-12 18:44:37,MONTGOMERY,600,1,Traffic,DISABLED VEHICLE -
56937,39.745533,-84.395256,MAIN ST; UPPER PROVIDENCE; 2016-05-07 @ 16:08:...,19475.0,Fire: VEHICLE FIRE,2016-05-07 16:08:25,UPPER PROVIDENCE,MAIN ST,1,Fire,VEHICLE FIRE
66737,30.333596,-95.595595,8931; MONTGOMERY; Station 311; 2016-06-02 @ 1...,77316.0,EMS: CARDIAC EMERGENCY,2016-06-02 13:31:21,MONTGOMERY,8931,1,EMS,CARDIAC EMERGENCY
88048,39.745533,-84.395256,MAIN ST; UPPER PROVIDENCE; 2016-07-27 @ 10:03:...,19475.0,Fire: GAS-ODOR/LEAK,2016-07-27 10:03:43,UPPER PROVIDENCE,MAIN ST,1,Fire,GAS-ODOR/LEAK
147405,39.745533,-84.395256,MAIN ST; UPPER PROVIDENCE; 2016-12-24 @ 02:43:...,19475.0,Fire: S/B AT HELICOPTER LANDING,2016-12-24 02:43:45,UPPER PROVIDENCE,MAIN ST,1,Fire,S/B AT HELICOPTER LANDING
...,...,...,...,...,...,...,...,...,...,...,...
659779,22.986757,87.854975,RAMP RT422 WB TO RT23 & RT422 WB; UPPER MERIO...,NaN,Traffic: VEHICLE ACCIDENT -,2020-07-18 22:24:14,UPPER MERION,RAMP RT422 WB TO RT23 & RT422 WB,1,Traffic,VEHICLE ACCIDENT -
659976,46.565316,-66.461916,PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB; W...,NaN,Traffic: DISABLED VEHICLE -,2020-07-19 15:03:20,WHITEMARSH,PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB,1,Traffic,DISABLED VEHICLE -
662023,46.565316,-66.461916,PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB; W...,NaN,Fire: VEHICLE ACCIDENT,2020-07-25 03:54:06,WHITEMARSH,PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB,1,Fire,VEHICLE ACCIDENT
662024,46.565316,-66.461916,PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB; W...,NaN,Traffic: VEHICLE ACCIDENT -,2020-07-25 03:53:59,WHITEMARSH,PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB,1,Traffic,VEHICLE ACCIDENT -


In [89]:
# Alternative method is
outliers = df[
    ~df['lat'].between(39.72, 42.27)|
    ~df['lng'].between(-80.52, -74.68)
]

print(outliers)

              lat        lng  \
25529   32.387090 -86.276106   
56937   39.745533 -84.395256   
66737   30.333596 -95.595595   
88048   39.745533 -84.395256   
147405  39.745533 -84.395256   
...           ...        ...   
659779  22.986757  87.854975   
659976  46.565316 -66.461916   
662023  46.565316 -66.461916   
662024  46.565316 -66.461916   
662323  46.565316 -66.461916   

                                                     desc      zip  \
25529           600  ; MONTGOMERY; 2016-02-12 @ 18:44:37;  36107.0   
56937   MAIN ST; UPPER PROVIDENCE; 2016-05-07 @ 16:08:...  19475.0   
66737   8931;  MONTGOMERY; Station 311; 2016-06-02 @ 1...  77316.0   
88048   MAIN ST; UPPER PROVIDENCE; 2016-07-27 @ 10:03:...  19475.0   
147405  MAIN ST; UPPER PROVIDENCE; 2016-12-24 @ 02:43:...  19475.0   
...                                                   ...      ...   
659779  RAMP RT422 WB TO RT23  & RT422 WB; UPPER MERIO...      NaN   
659976  PENNSYLVANIA TPKE & RAMP I476 NB TO I276 EB; W.

In [90]:
# Remove clearly wrong rows
df = df[(df['lat'].between(39.72, 42.27)) & (df['lng'].between(-80.52, -74.68))]

b. Check Abnormal TimeStamps

In [91]:
# check range
df['timeStamp'].min(), df['timeStamp'].max()

('2015-12-10 14:39:21', '2020-07-29 15:54:08')

### Step 4: Features Engineering

In [ ]:
# extract hour, day of week, and month from the timeStamp colum
df['hour'] = df['timeStamp'].dt.hour
df['day_of_week'] = df['timeStamp'].dt.day_name()
df['month'] = df['timeStamp'].dt.month

In [ ]:
# Visualize hourly trends
df.groupby('hour')['e'].count().plot()

In [ ]:
# Visualize Calls by Day of Week
df.groupby('day_of_week')['e'].count().plot(kind='bar') 

In [ ]:
# Visualize Calls by Township
df['twp'].value_counts().head(10).plot(kind='barh')

In [ ]:
# Visualize Monthly Trends
df.groupby('month')['e'].count().plot()
plt.show()

In [ ]:
# Plot a Heatmap (Day vs Hour)
import seaborn as sns


pivot = df.pivot_table(index='day_of_week', columns='hour', values='e', aggfunc='count')

sns.heatmap(pivot)

### Step 4: Encoding of Categorical Data

We are going to apply One'-Hot Encoding to the dept column

In [ ]:
dummies = pd.get_dummies(df.dept)
dummies = dummies.astype(int)
print(dummies)